Phi-3-mini-128k-instruct model bringup

ign-msati opened the pull request:
### Ticket
Link to Github Issue : #19416 

### Problem description
`microsoft/phi-3-mini-128k-instruct` model bringup on Tenstorrent Hardware (N300/N150)

### What's changed

- Added support for `microsoft/phi-3-mini-128k-instruct` using` tt-metal/models/tt_transformers`, reusing modules with minimal changes.
- Updated `RotarySetup()` class to support scaling using a scale-tensor for Phi3.
- Currenlty tested with N150 & N300 devices

### Supports  context length
- N150 : 32k
- N300 : 64k

### Accuracy Metrics
This configuration uses bfp8 for MLP (FF1+FF3), Attention weights and kv-cache. **Batch_size=1 and prefill_length is 128 tokens.**
| Device | Top-1 (%) | Top-5 (%) | Speed (t/s/u) | TTFT (ms) |
|--------|-----------|-----------|---------------|-----------|
| N150   | 94        | 99        | 40.8          | 96       |
| N300   | 94        | 99        | 55.7          | 65.5        |
### Performance Metrics
This configuration uses bfp4 MLP (FF1+FF3) and bfp8 for Attention weights and kv-cache. **Batch_size=1 (per DP group) and prefill_length is 128 tokens.**
| Device     | Top-1 (%) | Top-5 (%) | Speed (t/s/u) | TTFT (ms) |
|------------|-----------|-----------|---------------|-----------|
| N150       | 90        | 99        | 45.6          | 72.8        |
| N300       | 89        | 98        | 59.5          | 57.9        |

##  Long-context
This configuration uses bfp4 MLP (FF1+FF3) and bfp8 for Attention weights and kv-cache. **Batch_size=1 and token generation of 200.**
### Device : N300
| Prefill length | Device  | Speed (t/s/u) | TTFT (ms) |
|----------------|---------|---------------|-----------|
| 1k             | N300    | 54.2          | 274.8     |
| 2k             | N300    | 52.5          | 293.7     |
| 4k             | N300    | 48.5          | 951.7     |
| 8k             | N300    | 42.6          | 1820.8    |
| 16k            | N300    | 33.2          | 4129      |
| 32k            | N300    | 25.2          | 10736     |

### Device : N150
| Prefill length | Device  | Speed (t/s/u) | TTFT (ms) |
|----------------|---------|---------------|-----------|
| 1k             | N150    | 42.05         | 390       |
| 2k             | N150    | 40.0          | 414.2     |
| 4k             | N150    | 35.8          | 1401.7    |
| 8k             | N150    | 30            | 2835.9    |
| 16k            | N150    | 21.9          | 6723.4    |
 
 


### Checklist
- [ ] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] (For models and ops writers) [Single-card demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) CI passes (if applicable) See [recommended dev flow](https://github.com/tenstorrent/tt-metal/blob/main/models/MODEL_ADD.md#a-recommended-dev-flow-on-github-for-adding-new-models).
- [ ] New/Existing tests provide coverage for changes

ign-msati committed:
Phi-3-mini-128k-instruct model bringup

ign-msati committed:
performance optimization by ~1 t/s/u

jvasilje commented:
@jonathansuTT @ayerofieiev-tt please take a quick look. :)

uaydonat commented:
@yieldthought to be review the usage of tt-transformers

yieldthought commented:
Very cool!

Most models can be seamlessly integrated into `models/tt_transformers` so that one codebase can run a wide range of models specified only by changing the `HF_MODEL` environment variable to name one. I see you have added several additional files for Phi3 mini. Can you tell me why each is necessary vs integrating directly into the main code?

ign-msati committed:
simplify rope scale selection

ign-msati committed:
update readme

yieldthought commented:
Yes we'd really love to make tt-transformers general enough to run this model out of the box as well as Qwen, Llama, Mistral etc.

ign-msati commented:
> Yes we'd really love to make tt-transformers general enough to run this model out of the box as well as Qwen, Llama, Mistral etc.

Thanks for the confirmation, will update this PR  by integrating all the above changes directly into `/tt_transformers`.

ign-msati committed:
Integrate Phi-3-mini-128k-instruct to tt_transformers

yieldthought reviewed (approved):
Looks great now!

yieldthought commented on models/tt_transformers/tt/model_config.py:
```diff
@@ -1455,8 +1470,26 @@ def _set_params_from_dict(self, params, is_hf=False):
             self.rope_scaling_factor = rope_scaling_params.get("factor", None)
             self.orig_context_len = rope_scaling_params.get("original_max_position_embeddings", None)
+            self.rope_ext_scaling_tensor = None
+            if self.base_model_name == "Phi-3-mini-128k-instruct":
```
Does this need to be restricted to Phi-3-mini? It looks like it would be the right thing to do for any model that uses `longrope`, right?

ign-msati committed:
updated model config as per PR review comments

cguerreroTT commented:
Hi @yieldthought, just following up on this.  Can you please approve the workflows?

mtairum commented:
@ign-msati Can you rebase your branch to the latest main and fix any conflicts?
We can then run the workflows and merge if they are passing.

FIY @gwangTT

ign-msati committed:
rebasing model_config

ign-msati commented:
> @ign-msati Can you rebase your branch to the latest main and fix any conflicts? We can then run the workflows and merge if they are passing.
> 
> FIY @gwangTT

branch is rebased and ready for workflow

mtairum commented:
I've added the fork as a branch to main repo here:

https://github.com/tenstorrent/tt-metal/tree/mirror/ign-msati/ign_mayank/19416_phi3_mini_model_bringup

During this week I plan to rebase the branch again, add the Phi3-mini to our N150/N300 CI tests and run the github pipelines.

I'll update this thread as things progress. If all looks good we can merge at the end and the bounty will be complete 👍

ign-msati commented:
> I've added the Phi-3-mini weights to our CI, rebased and added the unit tests to Phi3. You can check my changes on the mirror branch (required for CI pipelines) here: [main...mirror/ign-msati/ign_mayank/19416_phi3_mini_model_bringup](https://github.com/tenstorrent/tt-metal/compare/main...mirror/ign-msati/ign_mayank/19416_phi3_mini_model_bringup)
> 
> However when running tests locally I see they failing.
> 
> @ign-msati Please make sure all the unit tests inside `models/tt_transformers/tests/test_ci_dispatch.py -k phi-3` are run correctly.
> 
> I see failures in the attention and decoder unit tests. Example logs below. After these are updated I'll mirror the changes to the branch in the main repo and run the CI pipelines.
> 
> * Attention-prefill fail log:
>   [log_test_attention_prefill.log](https://github.com/user-attachments/files/21233312/log_test_attention_prefill.log)
> * Decoder-prefill fail log:
>   [log_test_decoder_prefill.log](https://github.com/user-attachments/files/21233317/log_test_decoder_prefill.log)
> 
> One extra thing. Phi-3 requires a more recent version of HuggingFace transformers than the one TT-Metal current uses. We currently have an internal PR to bump TT-Metal transformers dependency, so hopefully that gets merged soon so we can run the Phi-3 CI pipelines.

@mtairum : Executed the unit test inside `models/tt_transformers/tests/test_ci_dispatch.py -k phi-3` in the PR branch as well in mirror branch [main...mirror/ign-msati/ign_mayank/19416_phi3_mini_model_bringup](https://github.com/tenstorrent/tt-metal/compare/main...mirror/ign-msati/ign_mayank/19416_phi3_mini_model_bringup) and test are executing successfully, 
For Phi-3-mini-128k-instruct, weights and checkpoints are loaded from https://huggingface.co/microsoft/Phi-3-mini-128k-instruct/tree/main.
Just wanted to check if you are using the same checkpoints as it was observed in the log file, rotary embedding is missing in attention module  <img width="1659" height="320" alt="image" src="https://github.com/user-attachments/assets/84ddfbed-ed29-484f-a642-b4640933d8d0" />

compared to the checkpoint used from the above [link ](https://huggingface.co/microsoft/Phi-3-mini-128k-instruct/tree/main.)

<img width="598" height="118" alt="image" src="https://github.com/user-attachments/assets/671b3947-e83d-43cf-9209-3964915850de" />

mtairum commented:
Just to be sure. 
How did you import the weights when you run the tests/demo.
Did you use `LLAMA_DIR=...` or `HF_MODEL=`?

ign-msati commented:
> Did you manually cloned the model?
> 
> And which version of transformers are you using?
yes manual cloned the model [link](https://huggingface.co/microsoft/Phi-3-mini-128k-instruct/tree/main.) 
 
Used transformer version (transformers==4.38.0) , actually rebuild from the your mirror branch and install all the libraries with the version present in https://github.com/tenstorrent/tt-metal/blob/mirror/ign-msati/ign_mayank/19416_phi3_mini_model_bringup/tt_metal/python_env/requirements-dev.txt

ign-msati commented:
its seems like the transformer version issue, let me know ideal version of transformer need to be considered, loading of rotary embedding can be done accordingly.

ign-msati committed:
added remote code flag for hf models

ign-msati commented:
@mtairum : PR is now working with `transformers == 4.53.0`

mtairum commented:
Thanks! I'll test it on my side 👍

fahmedchanTT commented:
@mtairum I see that the version bump PR is merged, is this unblocked ?

ign-msati commented:
@mtairum  Rebasing to the current main requires major changes in `rope.py`, unfortunately I lost access to instance in koyeb due to change in billing from August 1, could you clarify if there will be any charges apply in using koyeb instance if I resume my  instance.

ign-msati commented:
@mtairum  : raised the new [PR](https://github.com/tenstorrent/tt-metal/pull/27289)
